File: labs/class-01-wrangling/sampler.py
#!/usr/bin/env python3
"""
Understanding Uncertainty - Wrangling & robust statistics sampler
(standalone, stdlib only)
------------------------------------------------------------------
Four statistics computed from scratch on the real 250-listing used-car
sample: the mean and variance (which trust every observation equally),
and the median and quantile (which only care about rank, so a single
extreme value cannot move them far). This mirrors sampler.js and
viz.js's WR.* core line-for-line, and the assignment template
(wrangling.py) implements median() and quantile() the same way.

Usage:
  python3 sampler.py describe <price|mileage>          # mean/var/sd/median/IQR
  python3 sampler.py mean     <price|mileage>
  python3 sampler.py variance <price|mileage>
  python3 sampler.py median   <price|mileage>
  python3 sampler.py quantile <price|mileage> <frac>   # frac in [0, 1]

Reads the sample from data.json sitting next to this file.
"""
import sys, os, json, math


# ---- Statistics of a sample ----------------------------------
def mean(X):
    return sum(X) / len(X)


def variance(X):
    m = mean(X)
    return sum((x - m) ** 2 for x in X) / len(X)


def std(X):
    return math.sqrt(variance(X))


def median(X):
    values = sorted(X)
    n = len(values)
    mid = n // 2
    if mid != n / 2:
        return values[mid]
    return (values[mid - 1] + values[mid]) / 2


def quantile(X, frac):
    values = sorted(X)
    n = len(values)
    index = max(math.ceil(n * frac) - 1, 0)
    return values[index]


def arcsinh(x):
    return math.log(x + math.sqrt(x * x + 1))


# ---- CLI -------------------------------------------------------
def load(var):
    here = os.path.dirname(os.path.abspath(__file__))
    with open(os.path.join(here, "data.json")) as f:
        data = json.load(f)
    if var not in ("price", "mileage"):
        sys.stderr.write("variable must be 'price' or 'mileage'\n")
        sys.exit(1)
    return data[var]


def main():
    argv = sys.argv[1:]
    if not argv:
        sys.stderr.write("usage: python3 sampler.py [describe|mean|variance|median|quantile] <price|mileage> [frac]\n")
        sys.exit(1)
    cmd = argv[0]

    if cmd == "describe":
        X = load(argv[1])
        q25, q75 = quantile(X, 0.25), quantile(X, 0.75)
        sys.stderr.write(f"# n={len(X)}  variable={argv[1]}\n")
        print(f"mean:     {mean(X):.4f}")
        print(f"variance: {variance(X):.4f}")
        print(f"sd:       {std(X):.4f}")
        print(f"median:   {median(X):.4f}")
        print(f"q25:      {q25:.4f}")
        print(f"q75:      {q75:.4f}")
        print(f"IQR:      {q75 - q25:.4f}")
    elif cmd == "mean":
        print(f"{mean(load(argv[1])):.4f}")
    elif cmd == "variance":
        print(f"{variance(load(argv[1])):.4f}")
    elif cmd == "median":
        print(f"{median(load(argv[1])):.4f}")
    elif cmd == "quantile":
        X = load(argv[1])
        frac = float(argv[2])
        print(f"{quantile(X, frac):.4f}")
    else:
        sys.stderr.write("usage: python3 sampler.py [describe|mean|variance|median|quantile] <price|mileage> [frac]\n")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: labs/class-01-wrangling/sampler.js
#!/usr/bin/env node
/* Understanding Uncertainty - Wrangling & robust statistics sampler
 * (standalone, pure JS, no dependencies)
 * ------------------------------------------------------------------
 * Mirrors sampler.py and viz.js's WR.* core line-for-line: the same
 * mean, variance, median, and quantile on the same 250-listing sample.
 *
 * Usage (Node):
 *   node sampler.js describe <price|mileage>
 *   node sampler.js mean     <price|mileage>
 *   node sampler.js variance <price|mileage>
 *   node sampler.js median   <price|mileage>
 *   node sampler.js quantile <price|mileage> <frac>
 *
 * Reads the sample from data.json sitting next to this file.
 */
'use strict';
const fs = require('fs');
const path = require('path');

function mean(X) { return X.reduce((a, b) => a + b, 0) / X.length; }
function variance(X) {
  const m = mean(X);
  return X.reduce((a, b) => a + (b - m) * (b - m), 0) / X.length;
}
function std(X) { return Math.sqrt(variance(X)); }
function median(X) {
  const v = X.slice().sort((a, b) => a - b);
  const n = v.length, mid = Math.floor(n / 2);
  if (mid !== n / 2) return v[mid];
  return (v[mid - 1] + v[mid]) / 2;
}
function quantile(X, frac) {
  const v = X.slice().sort((a, b) => a - b);
  const n = v.length;
  const idx = Math.max(Math.ceil(n * frac) - 1, 0);
  return v[idx];
}
function fmt(x) { return x.toFixed(4); }

function load(varName) {
  const data = JSON.parse(fs.readFileSync(path.join(__dirname, 'data.json'), 'utf8'));
  if (varName !== 'price' && varName !== 'mileage') {
    process.stderr.write("variable must be 'price' or 'mileage'\n");
    process.exit(1);
  }
  return data[varName];
}

function main() {
  const argv = process.argv.slice(2);
  if (!argv.length) {
    process.stderr.write('usage: node sampler.js [describe|mean|variance|median|quantile] <price|mileage> [frac]\n');
    process.exit(1);
  }
  const cmd = argv[0];
  if (cmd === 'describe') {
    const X = load(argv[1]);
    const q25 = quantile(X, 0.25), q75 = quantile(X, 0.75);
    process.stderr.write(`# n=${X.length}  variable=${argv[1]}\n`);
    console.log('mean:     ' + fmt(mean(X)));
    console.log('variance: ' + fmt(variance(X)));
    console.log('sd:       ' + fmt(std(X)));
    console.log('median:   ' + fmt(median(X)));
    console.log('q25:      ' + fmt(q25));
    console.log('q75:      ' + fmt(q75));
    console.log('IQR:      ' + fmt(q75 - q25));
  } else if (cmd === 'mean') {
    console.log(fmt(mean(load(argv[1]))));
  } else if (cmd === 'variance') {
    console.log(fmt(variance(load(argv[1]))));
  } else if (cmd === 'median') {
    console.log(fmt(median(load(argv[1]))));
  } else if (cmd === 'quantile') {
    const X = load(argv[1]);
    console.log(fmt(quantile(X, parseFloat(argv[2]))));
  } else {
    process.stderr.write('usage: node sampler.js [describe|mean|variance|median|quantile] <price|mileage> [frac]\n');
    process.exit(1);
  }
}

main();
